File: C++/src/DoubleLinkedList.cpp
#include "../include/DoubleLinkedList.h"
#include <stdexcept>

DoubleLinkedList::DoubleLinkedList() : head(nullptr), tail(nullptr), size(0) {}

DoubleLinkedList::DoubleLinkedList(const DoubleLinkedList &other)
    : head(nullptr), tail(nullptr), size(0) {
  Node *current = other.head;
  while (current != nullptr) {
    pushBack(current->data);
    current = current->next;
  }
}

DoubleLinkedList::~DoubleLinkedList() { clear(); }

DoubleLinkedList &DoubleLinkedList::operator=(const DoubleLinkedList &other) {
  if (this != &other) {
    clear();
    Node *current = other.head;
    while (current != nullptr) {
      pushBack(current->data);
      current = current->next;
    }
  }
  return *this;
}

void DoubleLinkedList::pushFront(const std::string &value) {
  Node *newNode = new Node(value);
  if (head == nullptr) {
    head = tail = newNode;
  } else {
    newNode->next = head;
    head->prev = newNode;
    head = newNode;
  }
  size++;
}

void DoubleLinkedList::pushBack(const std::string &value) {
  Node *newNode = new Node(value);
  if (tail == nullptr) {
    head = tail = newNode;
  } else {
    newNode->prev = tail;
    tail->next = newNode;
    tail = newNode;
  }
  size++;
}

void DoubleLinkedList::insert(int index, const std::string &value) {
  if (index < 0 || index > size) {
    throw std::out_of_range("Index out of range");
  }
  if (index == 0) {
    pushFront(value);
    return;
  }
  if (index == size) {
    pushBack(value);
    return;
  }

  Node *newNode = new Node(value);
  Node *current = head;
  for (int i = 0; i < index; i++) {
    current = current->next;
  }

  newNode->next = current;
  newNode->prev = current->prev;
  current->prev->next = newNode;
  current->prev = newNode;
  size++;
}

std::string DoubleLinkedList::get(int index) const {
  if (index < 0 || index >= size) {
    throw std::out_of_range("Index out of range");
  }

  Node *current;
  if (index < size / 2) {
    current = head;
    for (int i = 0; i < index; i++) {
      current = current->next;
    }
  } else {
    current = tail;
    for (int i = size - 1; i > index; i--) {
      current = current->prev;
    }
  }
  return current->data;
}

void DoubleLinkedList::set(int index, const std::string &value) {
  if (index < 0 || index >= size) {
    throw std::out_of_range("Index out of range");
  }

  Node *current;
  if (index < size / 2) {
    current = head;
    for (int i = 0; i < index; i++) {
      current = current->next;
    }
  } else {
    current = tail;
    for (int i = size - 1; i > index; i--) {
      current = current->prev;
    }
  }
  current->data = value;
}

void DoubleLinkedList::popFront() {
  if (head == nullptr) {
    throw std::runtime_error("List is empty");
  }

  Node *temp = head;
  head = head->next;
  if (head != nullptr) {
    head->prev = nullptr;
  } else {
    tail = nullptr;
  }
  delete temp;
  size--;
}

void DoubleLinkedList::popBack() {
  if (tail == nullptr) {
    throw std::runtime_error("List is empty");
  }

  Node *temp = tail;
  tail = tail->prev;
  if (tail != nullptr) {
    tail->next = nullptr;
  } else {
    head = nullptr;
  }
  delete temp;
  size--;
}

void DoubleLinkedList::remove(int index) {
  if (index < 0 || index >= size) {
    throw std::out_of_range("Index out of range");
  }
  if (index == 0) {
    popFront();
    return;
  }
  if (index == size - 1) {
    popBack();
    return;
  }

  Node *current = head;
  for (int i = 0; i < index; i++) {
    current = current->next;
  }

  current->prev->next = current->next;
  current->next->prev = current->prev;
  delete current;
  size--;
}

bool DoubleLinkedList::removeByValue(const std::string &value) {
  Node *current = head;
  while (current != nullptr) {
    if (current->data == value) {
      if (current == head) {
        popFront();
      } else if (current == tail) {
        popBack();
      } else {
        current->prev->next = current->next;
        current->next->prev = current->prev;
        delete current;
        size--;
      }
      return true;
    }
    current = current->next;
  }
  return false;
}

void DoubleLinkedList::clear() {
  while (head != nullptr) {
    Node *temp = head;
    head = head->next;
    delete temp;
  }
  tail = nullptr;
  size = 0;
}

void DoubleLinkedList::serialize(std::ofstream &out) const {
  out.write(reinterpret_cast<const char *>(&size), sizeof(size));
  Node *current = head;
  while (current != nullptr) {
    int len = current->data.length();
    out.write(reinterpret_cast<const char *>(&len), sizeof(len));
    out.write(current->data.c_str(), len);
    current = current->next;
  }
}

void DoubleLinkedList::deserialize(std::ifstream &in) {
  clear();
  int newSize;
  in.read(reinterpret_cast<char *>(&newSize), sizeof(newSize));

  for (int i = 0; i < newSize; i++) {
    int len;
    in.read(reinterpret_cast<char *>(&len), sizeof(len));
    char *buffer = new char[len + 1];
    in.read(buffer, len);
    buffer[len] = '\0';
    pushBack(std::string(buffer));
    delete[] buffer;
  }
}

File: go/structs/doublelinkedlist/doublelinkedlist.go
package doublelinkedlist

import (
	"encoding/json"
	"fmt"
)

// Node represents a doubly linked list node
type Node struct {
	Data string
	Next *Node
	Prev *Node
}

// DoubleLinkedList represents a doubly linked list
type DoubleLinkedList struct {
	head *Node
	tail *Node
	size int
}

// New creates a new DoubleLinkedList
func New() *DoubleLinkedList {
	return &DoubleLinkedList{}
}

// PushFront adds an element to the front
func (l *DoubleLinkedList) PushFront(value string) {
	newNode := &Node{Data: value}

	if l.head == nil {
		l.head = newNode
		l.tail = newNode
	} else {
		newNode.Next = l.head
		l.head.Prev = newNode
		l.head = newNode
	}
	l.size++
}

// PushBack adds an element to the back
func (l *DoubleLinkedList) PushBack(value string) {
	newNode := &Node{Data: value}

	if l.tail == nil {
		l.head = newNode
		l.tail = newNode
	} else {
		newNode.Prev = l.tail
		l.tail.Next = newNode
		l.tail = newNode
	}
	l.size++
}

// PopFront removes the first element
func (l *DoubleLinkedList) PopFront() error {
	if l.head == nil {
		return fmt.Errorf("list is empty")
	}

	l.head = l.head.Next
	if l.head != nil {
		l.head.Prev = nil
	} else {
		l.tail = nil
	}
	l.size--
	return nil
}

// PopBack removes the last element
func (l *DoubleLinkedList) PopBack() error {
	if l.tail == nil {
		return fmt.Errorf("list is empty")
	}

	l.tail = l.tail.Prev
	if l.tail != nil {
		l.tail.Next = nil
	} else {
		l.head = nil
	}
	l.size--
	return nil
}

// Get returns element at index (optimized for bidirectional access)
func (l *DoubleLinkedList) Get(index int) (string, error) {
	if index < 0 || index >= l.size {
		return "", fmt.Errorf("index out of range: %d", index)
	}

	var current *Node
	if index < l.size/2 {
		current = l.head
		for i := 0; i < index; i++ {
			current = current.Next
		}
	} else {
		current = l.tail
		for i := l.size - 1; i > index; i-- {
			current = current.Prev
		}
	}
	return current.Data, nil
}

// Size returns the number of elements
func (l *DoubleLinkedList) Size() int {
	return l.size
}

// IsEmpty checks if list is empty
func (l *DoubleLinkedList) IsEmpty() bool {
	return l.size == 0
}

// Clear removes all elements
func (l *DoubleLinkedList) Clear() {
	l.head = nil
	l.tail = nil
	l.size = 0
}

// ToSlice returns all elements as slice
func (l *DoubleLinkedList) ToSlice() []string {
	result := make([]string, 0, l.size)
	current := l.head
	for current != nil {
		result = append(result, current.Data)
		current = current.Next
	}
	return result
}

// MarshalJSON implements json.Marshaler
func (l *DoubleLinkedList) MarshalJSON() ([]byte, error) {
	return json.Marshal(l.ToSlice())
}

// UnmarshalJSON implements json.Unmarshaler
func (l *DoubleLinkedList) UnmarshalJSON(data []byte) error {
	var items []string
	if err := json.Unmarshal(data, &items); err != nil {
		return err
	}

	l.Clear()
	for _, item := range items {
		l.PushBack(item)
	}
	return nil
}
